File: CS/WebSite/App_Code/XpoHelper.cs
using System;
using System.Collections.Generic;
using System.Web;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;
using System.Threading;

public class FileManagerDemoXpoHelper {
    public static Session GetNewSession() {
        return new Session(DataLayer);
    }

    readonly static object _lockObject = new object();
    static object _dataLayer;
    static IDataLayer DataLayer {
        get {
            if(_dataLayer == null) {
                lock(_lockObject) {
                    if(Thread.VolatileRead(ref _dataLayer) == null) {
                        Thread.VolatileWrite(ref _dataLayer, GetDataLayer());
                    }
                }
            }
            return (IDataLayer)_dataLayer;
        }
    }

    static IDataLayer GetDataLayer() {
        XpoDefault.Session = null;
        string conn = AccessConnectionProvider.GetConnectionString(@"|DataDirectory|\Data.mdb");
        XPDictionary dict = new DevExpress.Xpo.Metadata.ReflectionDictionary();
        IDataStore store = XpoDefault.GetConnectionProvider(conn, AutoCreateOption.SchemaAlreadyExists);
        dict.GetDataStoreSchema(typeof(PersistentObjects.ArtsEntity).Assembly);

        IDataLayer dl = new ThreadSafeDataLayer(dict, store);
        return dl;
    }
}

File: VB/WebSite/App_Code/XpoHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Web
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports System.Threading

Public Class FileManagerDemoXpoHelper
	Public Shared Function GetNewSession() As Session
		Return New Session(DataLayer)
	End Function

	Private ReadOnly Shared _lockObject As Object = New Object()
	Private Shared _dataLayer As Object
	Private Shared ReadOnly Property DataLayer() As IDataLayer
		Get
			If _dataLayer Is Nothing Then
				SyncLock _lockObject
					If Thread.VolatileRead(_dataLayer) Is Nothing Then
						Thread.VolatileWrite(_dataLayer, GetDataLayer())
					End If
				End SyncLock
			End If
			Return CType(_dataLayer, IDataLayer)
		End Get
	End Property

	Private Shared Function GetDataLayer() As IDataLayer
		XpoDefault.Session = Nothing
		Dim conn As String = AccessConnectionProvider.GetConnectionString("|DataDirectory|\Data.mdb")
		Dim dict As XPDictionary = New DevExpress.Xpo.Metadata.ReflectionDictionary()
		Dim store As IDataStore = XpoDefault.GetConnectionProvider(conn, AutoCreateOption.SchemaAlreadyExists)
		dict.GetDataStoreSchema(GetType(PersistentObjects.ArtsEntity).Assembly)

		Dim dl As IDataLayer = New ThreadSafeDataLayer(dict, store)
		Return dl
	End Function
End Class
